sklearn/metrics/_scorer.py

```diff
@@ -133,7 +132,28 @@ def __call__(self, estimator, *args, **kwargs):
                 **{name: Bunch(score=kwargs) for name in self._scorers}
             )
 
-        for name, scorer in self._scorers.items():
+        # To have the highest cache hit rate, we need to check if we have any
+        # _ThresholdScorer and choose a single response method based on the current
+        # estimator
+        def process_scorer(scorer):
+            if not isinstance(scorer, _ThresholdScorer):
+                return scorer
+            if scorer._response_method is None:
+                response_method = ("decision_function", "predict_proba")
+            else:
+                response_method = scorer._response_method
+            new_response_method = _check_response_method(
+                estimator, response_method
+            ).__name__
+
+            new_scorer = copy.copy(scorer)
+            new_scorer._response_method = new_response_method
+            return new_scorer
```

Comment:
I'm not sure (yet) why we're doing this here. This doesn't play nicely with scorers which have taken mutable objects such as arrays and lists and all, which they do as extra args.